server/auth/auth_test.go

```diff
@@ -0,0 +1,304 @@
+	c.Assert(errs.ErrRoleHasPermission.Equal(err), IsTrue)
+
+	err = m.AddPermission("reader", Permission{Resource: "region", Action: "update"})
+	c.Assert(err, IsNil)
+
+	role, err := m.GetRole("reader")
+	c.Assert(err, IsNil)
+	c.Assert(role.GetPermissions(), DeepEquals, []Permission{
+		{Resource: "region", Action: "get"},
+		{Resource: "region", Action: "list"},
+		{Resource: "store", Action: "get"},
+		{Resource: "store", Action: "list"},
+		{Resource: "region", Action: "update"},
+	})
+}
+
+func (s *testAuthSuite) testRemovePermission(c *C, m *roleManager) {
+	err := m.RemovePermission("reader", Permission{Resource: "region", Action: "update"})
+	c.Assert(err, NotNil)
+	c.Assert(errs.ErrRoleMissingPermission.Equal(err), IsTrue)
+
+	err = m.RemovePermission("reader", Permission{Resource: "region", Action: "get"})
+	c.Assert(err, IsNil)
+
+	role, err := m.GetRole("reader")
+	c.Assert(err, IsNil)
+	c.Assert(role.GetPermissions(), DeepEquals, []Permission{
+		{Resource: "store", Action: "list"},
+		{Resource: "region", Action: "list"},
+		{Resource: "store", Action: "get"},
+	})
+}
+
+func initKV(c *C, client *clientv3.Client, rootPath string) {
+	_, err := client.Delete(context.TODO(), rootPath, clientv3.WithPrefix())
+	c.Assert(err, IsNil)
+
+	rolePrefix := path.Join(rootPath, "roles")
+
+	roles := []struct {
+		Name        string `json:"name"`
+		Permissions []struct {
+			Resource string `json:"resource"`
+			Action   string `json:"action"`
+		} `json:"permissions"`
+	}{
+		{Name: "reader", Permissions: []struct {
+			Resource string `json:"resource"`
+			Action   string `json:"action"`
+		}{
+			{Resource: "region", Action: "get"},
+			{Resource: "region", Action: "list"},
+			{Resource: "store", Action: "get"},
+			{Resource: "store", Action: "list"},
+		}},
+		{Name: "writer", Permissions: []struct {
+			Resource string `json:"resource"`
+			Action   string `json:"action"`
+		}{
+			{Resource: "region", Action: "delete"},
+			{Resource: "region", Action: "update"},
+			{Resource: "store", Action: "delete"},
+			{Resource: "store", Action: "update"},
+		}},
+		{Name: "admin", Permissions: []struct {
+			Resource string `json:"resource"`
+			Action   string `json:"action"`
+		}{
+			{Resource: "region", Action: "delete"},
+			{Resource: "region", Action: "get"},
+			{Resource: "region", Action: "list"},
+			{Resource: "region", Action: "update"},
+			{Resource: "store", Action: "delete"},
+			{Resource: "store", Action: "get"},
+			{Resource: "store", Action: "list"},
+			{Resource: "store", Action: "update"},
+			{Resource: "users", Action: "delete"},
+			{Resource: "users", Action: "get"},
+			{Resource: "users", Action: "list"},
+			{Resource: "users", Action: "update"},
+		}},
+	}
+	for _, role := range roles {
+		value, err := json.Marshal(role)
+		c.Assert(err, IsNil)
+		_, err = client.Put(context.TODO(), path.Join(rolePrefix, role.Name), string(value))
+		c.Assert(err, IsNil)
+	}
+}
+
+func newTestSingleConfig() *embed.Config {
```

Comment:
Does `NewTestSingleConfig` in `server` package meet the requirement?

Reply:
Actually, it seems that etcd is not needed to run these tests. I'm replacing EtcdKV with MemoryKV.